// @wundergraph/wunderctl/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wunderctlBinaryPath = exports.binaryDir = exports.downloadURL = exports.wunderctlName = exports.cli = exports.uninstaller = exports.installer = void 0;
var installer_1 = require("./installer");
Object.defineProperty(exports, "installer", { enumerable: true, get: function () { return installer_1.installer; } });
var uninstaller_1 = require("./uninstaller");
Object.defineProperty(exports, "uninstaller", { enumerable: true, get: function () { return uninstaller_1.uninstaller; } });
var cli_1 = require("./cli");
Object.defineProperty(exports, "cli", { enumerable: true, get: function () { return cli_1.cli; } });
var binarypath_1 = require("./binarypath");
Object.defineProperty(exports, "wunderctlName", { enumerable: true, get: function () { return binarypath_1.wunderctlName; } });
Object.defineProperty(exports, "downloadURL", { enumerable: true, get: function () { return binarypath_1.downloadURL; } });
Object.defineProperty(exports, "binaryDir", { enumerable: true, get: function () { return binarypath_1.binaryDir; } });
Object.defineProperty(exports, "wunderctlBinaryPath", { enumerable: true, get: function () { return binarypath_1.wunderctlBinaryPath; } });
